pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

```diff
@@ -50,13 +62,530 @@ func New(_ runtime.Object, fh framework.FrameworkHandle) (framework.Plugin, erro
 
 // PostFilter invoked at the postFilter extension point.
 func (pl *DefaultPreemption) PostFilter(ctx context.Context, state *framework.CycleState, pod *v1.Pod, m framework.NodeToStatusMap) (*framework.PostFilterResult, *framework.Status) {
-	nnn, err := core.Preempt(ctx, pl.fh, state, pod, m)
+	nnn, err := Preempt(ctx, pl.fh, state, pod, m)
 	if err != nil {
 		return nil, framework.NewStatus(framework.Error, err.Error())
 	}
 	if nnn == "" {
 		return nil, framework.NewStatus(framework.Unschedulable)
-	} else {
-		return &framework.PostFilterResult{NominatedNodeName: nnn}, framework.NewStatus(framework.Success)
 	}
+	return &framework.PostFilterResult{NominatedNodeName: nnn}, framework.NewStatus(framework.Success)
+}
+
+// Preempt finds nodes with pods that can be preempted to make room for "pod" to
+// schedule. It chooses one of the nodes and preempts the pods on the node and
+// returns 1) the node, 2) the list of preempted pods if such a node is found,
+// 3) A list of pods whose nominated node name should be cleared, and 4) any
+// possible error.
+// Preempt does not update its snapshot. It uses the same snapshot used in the
+// scheduling cycle. This is to avoid a scenario where preempt finds feasible
+// nodes without preempting any pod. When there are many pending pods in the
+// scheduling queue a nominated pod will go back to the queue and behind
+// other pods with the same priority. The nominated pod prevents other pods from
+// using the nominated resources and the nominated pod could take a long time
+// before it is retried after many other pending pods.
+func Preempt(ctx context.Context, fh framework.FrameworkHandle, state *framework.CycleState, pod *v1.Pod, m framework.NodeToStatusMap) (string, error) {
```

Comment:
don't export?

Reply:
I think it depends on which kind of developers we care about:

- If the developers are following the scheduler framework, they don't need to call this function - as in that case, they just use the defaultpreemption plugin; otherwise, they built their "preempt" function using combined built-in preemption libraries and customized ones. 
- The other case is free-style scheduler implementation, they may want to call it directly as they're not using scheduler framework.

I'll evaluate whether this function, as well as its internal functions, needs to be made public, in the next PR.

Reply:
sounds good, until then I think we should just follow the normal golang best practices which tells us that this should be internal.

Reply:
Yes, making it private makes more sense. Let me change it in this PR later in case it's out of track.